File: Test/10_symphony_compare/test_symphony_compare.R
# CASSIA Test 10: Symphony Compare (R)
# =====================================
# Tests the symphonyCompare function via R package for multi-model
# cell type comparison with AI consensus building.
#
# Usage:
#     Rscript test_symphony_compare.R
#
# Functions tested:
# - symphonyCompare(): Orchestrate multiple AI models to compare cell types

# Get script directory (works with Rscript and source())
get_script_dir <- function() {
  args <- commandArgs(trailingOnly = FALSE)
  file_arg <- grep("--file=", args, value = TRUE)
  if (length(file_arg) > 0) {
    return(dirname(normalizePath(sub("--file=", "", file_arg), winslash = "/")))
  }
  for (i in sys.nframe():1) {
    if (!is.null(sys.frame(i)$ofile)) {
      return(dirname(normalizePath(sys.frame(i)$ofile, winslash = "/")))
    }
  }
  return(normalizePath(getwd(), winslash = "/"))
}
script_dir <- get_script_dir()

# Source shared utilities
source(file.path(script_dir, "..", "shared", "r", "test_utils.R"))
source(file.path(script_dir, "..", "shared", "r", "fixtures.R"))
source(file.path(script_dir, "..", "shared", "r", "result_manager.R"))
source(file.path(script_dir, "..", "shared", "r", "logging_manager.R"))

run_symphony_compare_test <- function() {
  print_test_header("10 - Symphony Compare (R)")

  # Load configuration
  config <- load_config()
  print_config_summary(config)

  # Setup API keys
  setup_api_keys()

  # Setup CASSIA R package
  tryCatch({
    setup_cassia()
    message("CASSIA R package loaded successfully")
  }, error = function(e) {
    message("Error loading CASSIA: ", e$message)
    return(FALSE)
  })

  # Get settings
  llm_config <- config$llm
  data_config <- config$data

  # Create results directory
  results <- create_results_dir("10_symphony_compare", get_test_mode())

  start_logging(results$logs)
  log_msg("Results will be saved to:", results$base)

  # Test parameters
  test_cluster <- "plasma cell"
  markers <- get_cluster_markers(test_cluster)
  marker_set <- paste(markers[1:15], collapse = ", ")

  # Cell types to compare
  celltypes <- c("Plasma cell", "B cell", "T cell")

  # Run tests
  start_time <- Sys.time()
  errors <- list()
  status <- "error"
  symphony_results <- list()

  tryCatch({
    log_msg("\n--- Test: symphonyCompare ---")
    log_msg("  Tissue:", data_config$tissue %||% "large intestine")
    log_msg("  Species:", data_config$species %||% "human")
    log_msg("  Cell types to compare:", paste(celltypes, collapse = ", "))
    log_msg("  Marker set:", substr(marker_set, 1, 50), "...")
    log_msg("  Model preset: budget (cost-effective for testing)")
    log_msg("  Discussion: Disabled (for faster testing)")

    result <- CASSIA::symphonyCompare(
      tissue = data_config$tissue %||% "large intestine",
      celltypes = celltypes,
      marker_set = marker_set,
      species = data_config$species %||% "human",
      model_preset = "budget",
      output_dir = results$outputs,
      output_basename = "symphony_test",
      enable_discussion = FALSE,
      max_discussion_rounds = 0,
      consensus_threshold = 0.6,
      generate_report = TRUE,
      verbose = TRUE
    )

    # Check result
    if (is.list(result)) {
      log_msg("\n--- Symphony Compare Results ---")
      log_msg("  Consensus:", result$consensus %||% "No consensus")
      log_msg("  Confidence:", sprintf("%.1f%%", (result$confidence %||% 0) * 100))
      log_msg("  CSV file:", basename(result$csv_file %||% "N/A"))
      log_msg("  HTML report:", basename(result$html_file %||% "N/A"))

      # Summary statistics
      summary_data <- result$summary
      if (!is.null(summary_data)) {
        log_msg("\n  Summary:")
        log_msg("    Models used:", summary_data$models_used %||% "N/A")
        log_msg("    Total rounds:", summary_data$total_rounds %||% "N/A")
        log_msg("    Consensus reached:", summary_data$consensus_reached %||% FALSE)
      }

      symphony_results <- list(
        consensus = result$consensus,
        confidence = result$confidence,
        csv_file = result$csv_file,
        html_file = result$html_file,
        summary = summary_data
      )

      # Validate results
      if (!is.null(result$csv_file) && file.exists(result$csv_file)) {
        status <- "passed"
      } else {
        status <- "failed"
        errors <- list("CSV file not created")
      }
    } else {
      status <- "failed"
      errors <- list("Unexpected result format")
    }

  }, error = function(e) {
    errors <<- list(e$message)
    status <<- "error"
    log_msg("\nError:", e$message)
  })

  duration <- as.numeric(difftime(Sys.time(), start_time, units = "secs"))

  # Save metadata and results
  metadata <- create_test_metadata(
    test_name = "symphony_compare",
    config = config,
    duration_seconds = duration,
    status = status,
    clusters_tested = as.list(celltypes),
    errors = errors
  )
  save_test_metadata(results$outputs, metadata)

  save_test_results(results$outputs, list(
    celltypes_compared = celltypes,
    marker_set = marker_set,
    model_preset = "budget",
    results = symphony_results
  ))

  # Print final result
  success <- status == "passed"
  print_test_result(success, paste("Duration:", round(duration, 2), "s"))


  stop_logging()
  return(success)
}

# Run test
success <- run_symphony_compare_test()
quit(status = if (success) 0 else 1)

File: Test/10_symphony_compare/test_symphony_compare.py
"""
CASSIA Test 10: Symphony Compare
=================================
Tests the symphonyCompare function for multi-model cell type comparison
with AI consensus building.

Usage:
    python test_symphony_compare.py

Functions tested:
- symphonyCompare(): Orchestrate multiple AI models to compare cell types
"""

import sys
import time
from pathlib import Path

# Add shared utilities to path
sys.path.insert(0, str(Path(__file__).parent.parent / "shared" / "python"))

from fixtures import get_cluster_markers
from test_utils import (
    setup_cassia_imports,
    load_config,
    setup_api_keys,
    print_test_header,
    print_test_result,
    print_config_summary,
    get_test_mode
)
from result_manager import (
    create_results_dir,
    save_test_metadata,
    save_test_results,
    create_test_metadata,
    setup_logging,
    cleanup_logging
)

# Setup CASSIA imports
setup_cassia_imports()


def run_symphony_compare_test():
    """Test Symphony Compare functionality."""
    print_test_header("10 - Symphony Compare")

    # Load configuration
    config = load_config()
    print_config_summary(config)

    # Setup API keys
    setup_api_keys()

    # Get LLM settings
    llm_config = config['llm']
    data_config = config['data']

    # Import CASSIA functions
    from CASSIA import symphonyCompare

    # Create results directory with organized structure
    results = create_results_dir("10_symphony_compare", get_test_mode())
    print(f"Results will be saved to: {results['base']}")

    # Setup logging to capture console output
    logging_ctx = setup_logging(results['logs'])

    # Test parameters - compare cell types with a marker set
    test_cluster = "plasma cell"
    markers = get_cluster_markers(test_cluster)
    marker_set = ", ".join(markers[:15])  # Use top 15 markers

    # Cell types to compare
    celltypes = ["Plasma cell", "B cell", "T cell"]

    # Run tests
    start_time = time.time()
    errors = []
    status = "error"
    symphony_results = {}

    try:
        print(f"\n--- Test: symphonyCompare ---")
        print(f"  Tissue: {data_config.get('tissue', 'large intestine')}")
        print(f"  Species: {data_config.get('species', 'human')}")
        print(f"  Cell types to compare: {', '.join(celltypes)}")
        print(f"  Marker set: {marker_set[:50]}...")
        print(f"  Model preset: budget (cost-effective for testing)")
        print(f"  Discussion: Disabled (for faster testing)")

        result = symphonyCompare(
            tissue=data_config.get('tissue', 'large intestine'),
            celltypes=celltypes,
            marker_set=marker_set,
            species=data_config.get('species', 'human'),
            model_preset="budget",  # Use budget preset for cost-effective testing
            output_dir=str(results['outputs']),
            output_basename="symphony_test",
            enable_discussion=False,  # Disable discussion for faster testing
            max_discussion_rounds=0,
            consensus_threshold=0.6,
            generate_report=True,
            verbose=True
        )

        # Check result structure
        if isinstance(result, dict):
            print(f"\n--- Symphony Compare Results ---")
            print(f"  Consensus: {result.get('consensus', 'No consensus')}")
            print(f"  Confidence: {result.get('confidence', 0):.1%}")
            print(f"  CSV file: {Path(result.get('csv_file', '')).name if result.get('csv_file') else 'N/A'}")
            print(f"  HTML report: {Path(result.get('html_file', '')).name if result.get('html_file') else 'N/A'}")

            # Summary statistics
            summary = result.get('summary', {})
            if summary:
                print(f"\n  Summary:")
                print(f"    Models used: {summary.get('models_used', 'N/A')}")
                print(f"    Total rounds: {summary.get('total_rounds', 'N/A')}")
                print(f"    Consensus reached: {summary.get('consensus_reached', False)}")

            # Cell type scores
            celltype_scores = summary.get('celltype_scores', {})
            if celltype_scores:
                print(f"\n  Cell Type Scores:")
                for ct, scores in celltype_scores.items():
                    print(f"    {ct}: mean={scores.get('mean', 0):.1f}, range=[{scores.get('min', 0):.1f}-{scores.get('max', 0):.1f}]")

            symphony_results = {
                'consensus': result.get('consensus'),
                'confidence': result.get('confidence'),
                'csv_file': result.get('csv_file'),
                'html_file': result.get('html_file'),
                'summary': summary,
                'num_results': len(result.get('results', []))
            }

            # Validate results
            if result.get('csv_file') and Path(result['csv_file']).exists():
                status = "passed"
            else:
                status = "failed"
                errors.append("CSV file not created or consensus not reached")
        else:
            status = "failed"
            errors.append("Unexpected result format")

    except Exception as e:
        errors.append(str(e))
        status = "error"
        print(f"\nError: {e}")
        import traceback
        traceback.print_exc()

    duration = time.time() - start_time

    # Save metadata and results
    metadata = create_test_metadata(
        test_name="symphony_compare",
        config=config,
        duration_seconds=duration,
        status=status,
        clusters_tested=celltypes,
        errors=errors
    )
    save_test_metadata(results['outputs'], metadata)

    save_test_results(results['outputs'], {
        "celltypes_compared": celltypes,
        "marker_set": marker_set,
        "model_preset": "budget",
        "results": symphony_results
    })

    # Print final result
    success = status == "passed"
    print_test_result(success, f"Duration: {duration:.2f}s")

    # Cleanup logging
    cleanup_logging(logging_ctx)

    return success


if __name__ == "__main__":
    success = run_symphony_compare_test()
    sys.exit(0 if success else 1)
